shared/src/androidMain/kotlin/io/customer/shared/tracking/queue/QueueTimer.kt

```diff
@@ -0,0 +1,102 @@
+    override val jobExecutor: JobExecutor,
+) : QueueTimer, JobDispatcher {
+    @Volatile
+    private var countdownTimer: CountDownTimer? = null
+    private val timerAlreadyScheduled: Boolean
+        get() = countdownTimer != null
+
+    override fun onCoroutineFailed(exception: Throwable) {
+        countdownTimer = null
+        logger.error(message = "Timer failed with error ${exception.message}")
+    }
+
+    override fun schedule(force: Boolean, duration: TimeUnit.Seconds, block: () -> Unit) {
+        synchronized(this) {
+            if (!force && timerAlreadyScheduled) {
+                log("already scheduled to run. Skipping request.")
+            }
+            scheduleAndCancelPrevious(duration = duration, block = block)
+        }
+    }
+
+    private fun scheduleAndCancelPrevious(duration: TimeUnit, block: () -> Unit) {
+        // Must create and start timer on the main UI thread or Android will throw an exception
+        // saying the current thread doesn't have a Looper.
+        // Because we are starting a new coroutine, there is a chance that there could be a delay
+        // in starting the timer. This is OK because this function is designed to be async anyway
+        // so the logic from the caller has not changed.
+        runOnMain(
+            onFailure = { ex ->
+                unsafeCancel()
+                logger.error(message = "Timer cannot be started, reason: ${ex.message}")
+            },
+        ) {
+            synchronized(this) {
+                unsafeCancel()
+
+                log("making a timer for $duration")
+
+                val millisInFuture = duration.toMillis()
+                // Since we only need to listen at completion, using the same duration for
+                // countdown interval
+                countdownTimer = object : CountDownTimer(millisInFuture, millisInFuture) {
+                    override fun onTick(millisUntilFinished: Long) {}
+                    override fun onFinish() {
+                        // reset timer before calling block as block might be synchronous and if it
+                        // tries to start a new timer, it will not succeed because we need to reset
+                        // the timer.
+                        timerDone()
+                        block()
+                    }
+                }.start()
+            }
+        }
+    }
+
+    private fun timerDone() {
+        synchronized(this) {
+            countdownTimer = null
+
+            log("timer is done! It's been reset")
+        }
+    }
+
+    override fun cancel() {
+        synchronized(this) {
+            log("timer is being cancelled")
+            unsafeCancel()
+        }
+    }
+
+    // cancel without having a mutex lock. Call within a synchronized{} block
+    private fun unsafeCancel() {
+        countdownTimer?.cancel()
+        countdownTimer = null
+    }
+
+    private fun log(message: String) {
+        logger.debug("QueueTimer $message")
```

Comment:
The iOS timer found some value in [adding an identifier](https://github.com/customerio/customerio-ios/blob/efe11854c5477ab6939f05c514a33b6210078968/Sources/Common/Util/Timer.swift#L19) to the log statements that identify a Timer instance from another. 

So, instead of getting logs like this:

```
Timer is being cancelled
```

you get:
```
Timer wofofoo is being cancelled 
```

There were some bugs that were really hard to debug in the past where there were multiple Timer instances being created and executed and we couldn't tell them apart. 

I recommend we add that to the KMM Timer, too.

Reply:
I think we might not have multiple Timer instances in KMM. But I still feel it can some add value, so I'll update this.

Reply:
I think we should add a comment to [this property](https://github.com/customerio/mobile-shared/pull/5/files#diff-88f1dda25027df8a87278e7353226c5d4aba325943981b4a92b296c555e523f9R29) explaining why we need it. I can see us deleting it in the future or wondering why we added it in the first place.

Reply:
Sure, will do

Reply:
Added comment